File: src/test/java/io/github/robertomike/baradum/filters/CustomFilterTest.java
package io.github.robertomike.baradum.filters;

import io.github.robertomike.baradum.Baradum;
import io.github.robertomike.baradum.configs.CustomTest;
import io.github.robertomike.baradum.configs.ParameterRequest;
import io.github.robertomike.baradum.models.User;
import io.github.robertomike.hefesto.builders.Hefesto;

import static org.mockito.Mockito.verify;

@CustomTest
public class CustomFilterTest {
    @ParameterRequest(key = "id", value = "1")
    void allowCustomFilter(Hefesto<User> hefesto) {
        Baradum.make(User.class)
                .allowedFilters(new CustomFilter("id", (builder, value) -> builder.where("name", value)))
                .get();

        verify(hefesto).where("name", "1");
    }
}

File: baradum-core/src/test/kotlin/io/github/robertomike/baradum/core/filters/CustomFilterTest.kt
package io.github.robertomike.baradum.core.filters

import io.github.robertomike.baradum.core.enums.BaradumOperator
import io.github.robertomike.baradum.core.interfaces.QueryBuilder
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock

/**
 * Tests for the generic, backend-agnostic CustomFilter in baradum-core - gives QueryDSL (and any
 * future backend) the same lambda-filter convenience baradum-hefesto's CustomFilter offers.
 */
class CustomFilterTest {

    @Test
    fun `filterByParam invokes the consumer with the query builder and value`() {
        val mockQueryBuilder = mock<QueryBuilder<Any>>()
        var capturedValue: String? = null

        val filter = CustomFilter<QueryBuilder<Any>>("status") { query, value ->
            capturedValue = value
            query.where("status", BaradumOperator.EQUAL, value)
        }

        filter.filterByParam(mockQueryBuilder, "premium")

        assertEquals("premium", capturedValue)
        org.mockito.kotlin.verify(mockQueryBuilder).where("status", BaradumOperator.EQUAL, "premium")
    }

    @Test
    fun `param and internalName both equal the given param name`() {
        val filter = CustomFilter<QueryBuilder<Any>>("myParam") { _, _ -> }

        assertEquals("myParam", filter.param)
        assertEquals("myParam", filter.internalName)
    }
}
